modules/social_features/social_group/src/Element/SocialGroupEntityAutocomplete.php

```diff
@@ -61,9 +61,30 @@ public static function validateEntityAutocomplete(array &$element, FormStateInte
         ];
 
         $account = User::load($match);
+        if ($group !== NULL && $group->hasField('field_flexible_group_visibility')) {
+          $group_visibilities = $group->get('field_flexible_group_visibility')
+            ->getValue();
+          $user_has_role = FALSE;
+          // Loop through group visibility and check if user has that role.
+          foreach ($group_visibilities as $group_visibility) {
+            if ($account !== NULL && $account->hasRole($group_visibility['value'])) {
+              $user_has_role = TRUE;
+              break;
+            }
+          }
+          // Add error if user doesn't have required role.
+          if ($account !== NULL && !$user_has_role) {
+            $message = \Drupal::translation()->translate(
+              "@username doesn't have the required role. You can't add them. Please contact your community manager.",
+              ['@username' => $account->getDisplayName()]
+            );
+            $form_state->setError($element, $message);
+            return;
+          }
+        }
         // User is already a member, add it to an array for the Form element
         // to render an error after all checks are gone.
-        if ($group->getMember($account)) {
+        if ($group !== NULL && $account !== NULL && $group->getMember($account)) {
           $duplicated_values[] = $account->getDisplayName();
         }
```

Comment:
It could be a good idea to pull the `if ($group !== NULL)` clause out of these two if statements. That way we can also add a comment explaining what the condition means, for example.

```
// We expect to have a group for the group entity autocomplete. However, in case
//  this is used outside of a group context we can still act as a normal user
// autocomplete. In that case it makes no sense to do group based validation since
// we know that a user isn't a member of a group we don't have.
```

Reply:
Couldn't we also check if the group is NULL and throw an exception as your suggestions for User? Because if we don't have a group, it doesn't make sense to further execute this validation function.